File: src/main/java/com/learning/leetcode/recursive/medium/WordSearch.java
package com.learning.leetcode.recursive.medium;

import java.util.Arrays;

public class WordSearch {

    public static void main(String[] args) {
        char[][] board = {{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
        searchWord(board, "ABCCED");
        searchWord(board, "FCEE");
    }

    private static void searchWord(char[][] board, String word){
        int sr = -1;
        int sc = -1;
        boolean wordFound = false;
        for(int i=0; i< board.length; i++){
            for(int j=0; j < board[0].length; j++){
                if(board[i][j] == word.charAt(0)){
                    //Search recursive function is called from every row/column
                    if(search(i,j,board,word,0)){
                        wordFound = true;
                        sr = i;
                        sc = j;
                        break;
                    }
                }
            }
        }
        for(char[] row: board){
            System.out.println(Arrays.toString(row));
        }
        if (wordFound) {
            System.out.println("Word " + word + " found at " + sr + "," + sc);
        } else {
            System.out.println("Word "+ word  + " not found");
        }
    }

    private static boolean search(int sr, int sc, char[][] board, String word, int wordIndex){
        if(wordIndex == word.length()) {
            return true;
        }
        if ( sr < 0 || sc < 0 || sr == board.length || sc == board[0].length || board[sr][sc] != word.charAt(wordIndex) ) {
            return false;
        }
        char ch = board[sr][sc];
        board[sr][sc] = '#';
        boolean op1 = search(sr+1, sc, board, word, wordIndex+1);
        boolean op2 = search(sr, sc+1, board, word, wordIndex+1);
        boolean op3 = search(sr-1, sc, board, word, wordIndex+1);
        boolean op4 = search(sr, sc-1, board, word, wordIndex+1);
        board[sr][sc] = ch;

        return op1 || op2 || op3 || op4;

    }
}

File: src/main/java/com/learning/leetcode/arrays/medium/WordSearch.py
from typing import List


def exist(board: List[List[str]], word: str) -> bool:
    for i in range(len(board)):
        for j in range(len(board[0])):
            if board[i][j] == word[0]:
                print("search starts at " + str(i) + "," + str(j))
                if search(board, word, 0, i, j):
                    return True
    return False


def search(board: List[List[str]], word: str, wordIndex: int, sr: int, sc: int) -> bool:
    if wordIndex == len(word):
        print("search ends at " + str(sr) + "," + str(sc))
        return True
    if sr < 0 or sc < 0 or sr == len(board) or sc == len(board[0]) or word[wordIndex] != board[sr][sc]:
        return False
    ch = board[sr][sc]
    board[sr][sc] = '#'
    op1 = search(board, word, wordIndex + 1, sr + 1, sc)
    op2 = search(board, word, wordIndex + 1, sr, sc + 1)
    op3 = search(board, word, wordIndex + 1, sr - 1, sc)
    op4 = search(board, word, wordIndex + 1, sr, sc - 1)
    board[sr][sc] = ch
    return op1 or op2 or op3 or op4


def main():
    board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
    word = "SEE"
    res = exist(board, word)
    for row in board:
        print(row)
    print("Word %s exists: %s" % (word, str(res)))
    board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
    word = "ABCCED"
    res = exist(board, word)
    for row in board:
        print(row)
    print("Word %s exists: %s" % (word, str(res)))
    board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
    word = "ABCB"
    res = exist(board, word)
    for row in board:
        print(row)
    print("Word %s exists: %s" % (word, str(res)))


main()
